File: src/math/segment.js
// @flow

import Point from './point';

export default class Segment {
  startPoint: Point;
  endPoint: Point;

  constructor(startPoint: Point, endPoint: Point) {
    this.startPoint = startPoint;
    this.endPoint = endPoint;
  }

  set setStartPoint(value: Point) {
    this.startPoint = value;
  }

  set setEndPoint(value: Point) {
    this.endPoint = value;
  }

  get abscissaLength(): number {
    return Math.abs(this.endPoint.x - this.startPoint.x);
  }

  get abscissaDirection(): number {
    return Math.sign(this.endPoint.x - this.startPoint.x);
  }

  get ordinateLength(): number {
    return Math.abs(this.endPoint.y - this.startPoint.y);
  }

  get ordinateDirection(): number {
    return Math.sign(this.endPoint.y - this.startPoint.y);
  }

  get longProjectionOnAxis(): number {
    return this.abscissaLength >= this.ordinateLength
      ? this.abscissaLength
      : this.ordinateLength;
  }
}

File: lib/math/segment.dart
import 'dart:math';

import 'Package:GraphicsApp/math/point.dart';


class Segment {
  Point2D _start;
  Point2D _end;

  Segment(this._start, this._end) {}

  Point2D get start => _start;

  void set start(Point2D point) {
    this._start = point;
  }

  Point2D get end => _end;

  void set end(Point2D point) {
    this._end = point;
  }

  num get abscissaLength {
    return (end.x - start.x).abs();
  }

  num get abscissaDirection {
    return (end.x - start.x).sign;
  }

  num get ordinateLength {
    return (end.y - start.y).abs();
  }

  num get ordinateDirection {
    return (end.y - start.y).sign;
  }

  num get longProjectionOnAxis {
    return abscissaLength >= ordinateLength ? abscissaLength : ordinateLength;
  }

  num get length {
    return sqrt(pow(end.x - start.x, 2) + pow(end.y - start.y, 2));
  }

  bool isPoint() {
    return end == start;
  }

  String toString() {
    return '${start}/${end}';
  }
}
